resources/web/docs_js/components/console_widget.jsx

```diff
@@ -69,23 +69,80 @@ export const ConsoleForm = connect((state, props) =>
         `${props.setting}_curl_password`], state.settings)
 , {saveSettings})(_ConsoleForm);
 
+const alternativePrettyName = rawName => {
+  switch(rawName) {
+    case 'console': return 'Console';
+    case 'csharp': return 'C#';
+    case 'js': return 'JavaScript';
+    case 'php': return 'PHP';
+    default: return rawName;
+  }
+};
+
+const alternativeChoice = rawName => {
+  return <option value={rawName}>{alternativePrettyName(rawName)}</option>;
+};
+
+const alternativePicker = props => {
+  if (!props.alternatives) {
+    return;
+  }
+  const consoleAlternatives = props.alternatives.console;
+  if (!consoleAlternatives) {
+    return;
+  }
+
+  const items = [];
+  let sawChoice = 'console' === props.consoleAlternative;
+  items.push(alternativeChoice('console'));
+  for (const name of Object.keys(consoleAlternatives)) {
+    sawChoice |= name === props.consoleAlternative;
+    items.push(alternativeChoice(name));
+  }
+
+  /* If value isn't in the list then *make* it and we'll render our standard
+   * "there no example for this language" option. This prevents us from
+   * squashing preferences that users set. */
+  if (!sawChoice) {
+    items.push(alternativeChoice(props.consoleAlternative));
+  }
+  // TODO we shouldn't change these drop downs after the first time they are rendered. The extra choice should stay while you stay on the page. Maybe we can get away with rendering this once on page load and never subscribing again?
+
+  // TODO add the "message" bubble to the warning.
+  // TODO prevent "jumping" when the size of the snippets isn't the same
+  return <div className="AlternativePicker u-space-between">
+    <select className="AlternativePicker-select"
+            value={props.consoleAlternative}
+            onChange={(e) => props.saveSettings({consoleAlternative: e.target.value})}>
+      {items}
+    </select>
+    <div className="AlternativePicker-warning" />
+  </div>;
+};
+
+// ConsoleWidget isn't quite the right name for this any more....
 export const ConsoleWidget = props => {
   const modalAction = () => props.openModal(ConsoleForm, {setting: props.setting, url_label: props.url_label});
-  return <div>
-    <a className="sense_widget copy_as_curl"
-       onClick={e => props.copyAsCurl({isKibana: props.isKibana, consoleText: props.consoleText, setting: props.setting})}>
-      {props.langStrings('Copy as cURL')}
-    </a>
-    {props.view_in_text &&
-      <a className="view_in_link"
-         target="console"
-         title={props.langStrings(props.view_in_text)}
-         href={`${props[props.setting + "_url"]}?load_from=${props.baseUrl}${props.snippet}`}>{props.langStrings(props.view_in_text)}</a>
-    }
-    <a className="console_settings" onClick={modalAction} title={props.langStrings(props.configure_text)}>&nbsp;</a>
+  return <div className="u-space-between">
+    {alternativePicker(props) || <div/>}
```

Comment:
Since it's a component we should 'render' it `<AlternativePicker {...props} />`

Reply:
and maybe it can return an empty div instead of the or here.

Reply:
I would actually change this to `<AlternativePicker />`, remove ` {alternatives: state.alternatives}` from map state to props below, and properly connect the AlternativePicker to grab `state.alternatives` and `state.settings.consoleAlternative`. That will move the concerns to the component that cares about it, and it's a good practice to get in the habit of.